### PR TITLE
feat: Set macOS notification close button title

### PR DESCRIPTION
**Pull request by sethlu**
🍎 This addition allows developers to set custom & localized title for macOS notifications.

Resolves: https://github.com/electron/electron/issues/11594

---

With the following config:

- `NSUserNotificationAlertStyle` set to `alert` in `Info.plist` of the app bundle
- Probably check if alerts are enabled for the app in System Preferences (not only banners, otherwise the close button won't show)

Using the following script:

```javascript
// This doesn't produce a window...
// But should be enough to produce a native notification
const electron = require('electron')
electron.app.on('ready', function (event) {

  let n = new electron.Notification({
    title: 'Notification',
    closeButtonText: 'Dismiss 😀', // It does support some emoji
  })

  n.show()

})
```

Screenshot:

screen shot 2018-01-16 at 11 26 20 am

### CODE REVIEW

**Comment by codebytere** [comment]
@sethlu this is 💯 , dope work! gonna wait on that last build then i'll merge 😁

**Comment by alexeykuzmin** [comment]
```js
  let n = new electron.Notification({
    title: 'Notification',
    closeButtonText: 'Dismiss 😀', // It does support some emoji
  })
```

@sethlu Can you please add a test that `n.closeButtonText` value is equal to the provided text?
And another test when `closeButtonText` is set to an empty string.

**Comment by sethlu** [comment]
@alexeykuzmin Thanks for the reviewing! This change set introduces a new member to the existing `NotificationOptions` struct. On platforms other than macOS, this `closeButtonText` will not be checked while interacting with the system API calls. However, I could wrap some of the macOS-only properties for optimization? (This hasn't been done yet however, probably due to it being an experimental feature? On that I also can't seem to find the notifications under the spec folder?)

File: atom/browser/api/atom_api_notification.cc [hunk 1]
```diff
@@ -214,23 +179,28 @@ bool Notification::IsSupported() {
 void Notification::BuildPrototype(v8::Isolate* isolate,
                                   v8::Local<v8::FunctionTemplate> prototype) {
   prototype->SetClassName(mate::StringToV8(isolate, "Notification"));
-  mate::ObjectTemplateBuilder(isolate, prototype->PrototypeTemplate())
-      .MakeDestroyable()
+  auto _(mate::ObjectTemplateBuilder(isolate, prototype->PrototypeTemplate()));
```
Comment (sethlu) [inline atom/browser/api/atom_api_notification.cc]: Not sure of a better way to do this while keeping the items in order?

**Comment by alexeykuzmin** [comment]
@sethlu 

> On platforms other than macOS, this closeButtonText will not be checked while interacting with the system API calls. However, I could wrap some of the macOS-only properties for optimization?

I wasn't thinking about optimization, but about explicitly saying "these parts of code are Mac-specific".

> I also can't seem to find the notifications under the spec folder

You are more than welcome to create a new file )
You can skip `.closeButtonText` tests on all platforms other than Mac by calling `this.skip()` in a `before` hook of a `describe`.

**Comment by sethlu** [comment]
@alexeykuzmin

> I wasn't thinking about optimization, but about explicitly saying "these parts of code are Mac-specific".

Would cd01eec address this? The feature's intended for macOS and previous additions to the Notifications API aren't so sensitive with the platform settings.

> You are more than welcome to create a new file )

I'll probably address that in a separate PR since the entire Notification submodule needs its tests--probably only checking if the values are passed correctly to the electron lib?

**Comment by alexeykuzmin** [comment]
@MarshallOfSound @sethlu 

> the` atom_api_notification` file could be considered a "proxy" for the internal `brightray` notification API. It shouldn't be aware what OS does what / supports rather just pass a complete `NotificationOptions` object to the API and let the OS specific stuff in `brightray` handle the options as they see fit

`atom_api_notification` is a user-facing interface, and it should match the documentation.
`brightray` is an implementation behind that interface, probably platform dependent.
We _can_ have some stuff implemented in `brightray` without an API available for users,
but we _can't_ have an API without implementation. Like if some API is not available on a certain platform, the method shouldn't be noop, it should not exist at all.

That's why I think that we _should_ have OS checks in the `atom_api_notification`.
I don't really care if some methods will end up in a separate file or not.

**Comment by MarshallOfSound** [comment]
> Like if some API method is not available on a certain platform, the method shouldn't be noop, it should not exist at all.

This goes against how most of our current API's work afaik, OS specific getters and setters exist most of the time on all platforms but only take affect on some.  This is so that you can write JS code for all platforms without doing platform checks.  I.e. I as a user shouldn't have to do a `if (process.platform === 'darwin')` check just to see if I can get/set the `hasReply` property.  We should make writing cross-platform as easy for users as possible (that's kind of our thing 😄 )

**Comment by sethlu** [comment]
@MarshallOfSound @alexeykuzmin I've just reverted the platform-specific code changes.
Let me know how it looks.

**Comment by alexeykuzmin** [comment]
@sethlu Looks good to me. Already approved.

**Comment by sethlu** [comment]
@MarshallOfSound @alexeykuzmin I've been a little busy lately. But just got to add some test cases to the existing Notification API.
